File: ruby/y2025/day08.rb
#!/usr/bin/env ruby
#
# Playground

require_relative '../day'
require_relative '../point'

class Day08 < Day
  def do_part1(lines)
    coords = parse_3d(lines)
    coords_and_dists = pairs_and_distances_between(coords)
                       .sort_by! { |pad| pad[2] }
    create_circuits(coords_and_dists, @testing ? 10 : 1000, [])
      .sort_by(&:length)
      .reverse
      .take(3)
      .reduce(1) { |product_of_lengths, circuit| product_of_lengths * circuit.length }
  end

  def do_part2(lines)
    coords = parse_3d(lines)
    coords_and_dists = pairs_and_distances_between(coords)
                       .sort_by! { |pad| pad[2] }
    c0, c1 = create_one_circuit(coords_and_dists, coords.length, [])
    c0.x * c1.x
  end

  private

  # Recursively connects coords and add to the proper circuits, creating and
  # merging circuits when necessary. Returns the resulting circuits after
  # `num_iters`.
  def create_circuits(coords_and_dists, num_iters, circuits)
    return circuits if num_iters == 0

    c0, c1 = coords_and_dists[0][..1]
    circuits = add_coords_to_circuits(c0, c1, circuits)
    create_circuits(coords_and_dists[1..], num_iters - 1, circuits)
  end

  # Recursively connects coords and add to the proper circuits, creating and
  # merging circuits when necessary. When we finally create one circuit
  # containing all the coordinates, return the last two coords that made
  # that happen.
  def create_one_circuit(coords_and_dists, num_coords, circuits)
    c0, c1 = coords_and_dists[0][..1]
    circuits = add_coords_to_circuits(c0, c1, circuits)
    return [c0, c1] if circuits.length == 1 && circuits[0].size == num_coords

    create_one_circuit(coords_and_dists[1..], num_coords, circuits)
  end

  # Given two coordinates and all existing circuits, connect them and update
  # circuits, returning the new list of circuits.
  def add_coords_to_circuits(c0, c1, circuits)
    c0_circuit = circuits.detect { |circuit| circuit.include?(c0) }
    c1_circuit = circuits.detect { |circuit| circuit.include?(c1) }

    if c0_circuit.nil? && c1_circuit.nil? # create a new circuit
      circuits << Set.new([c0, c1])
    elsif c0_circuit.nil?       # add c0 to c1 circuit
      c1_circuit << c0
    elsif c1_circuit.nil?       # add c1 to c0 circuit
      c0_circuit << c1
    elsif c0_circuit == c1_circuit # do nothing
      # nop
    else # two diff circuits; merge them
      c0_circuit.merge(c1_circuit)
      circuits.delete(c1_circuit)
    end
    circuits
  end

  def parse_3d(lines)
    lines.map do |line|
      Point.new(*line.split(',').map(&:to_i))
    end
  end

  # Returns an array of [c0, c1, squared-distance].
  def pairs_and_distances_between(coords)
    coords.combination(2).map do |c0, c1|
      [c0, c1, c0.distance_squared(c1)]
    end
  end
end

if __FILE__ == $PROGRAM_NAME
  require_relative '../aoc'

  aoc(__FILE__)
end

File: elixir/lib/y2025/day08.ex
# ================ Playground ================

defmodule Y2025.Day08 do
  alias Common.Enum, as: CE

  def part1(ctx, lines) do
    coords = parse_3d(lines)

    coords_and_dists =
      coords
      |> pairs_and_distances_between
      |> Enum.sort_by(fn {_, _, dist} -> dist end)

    coords_and_dists
    |> create_circuits(if(ctx.test, do: 10, else: 1000), MapSet.new())
    |> Enum.sort_by(&MapSet.size/1, :desc)
    |> Enum.take(3)
    |> Enum.map(&MapSet.size/1)
    |> Enum.product()
  end

  def part2(_ctx, lines) do
    coords = parse_3d(lines)

    coords_and_dists =
      coords
      |> pairs_and_distances_between
      |> Enum.sort_by(fn {_, _, dist} -> dist end)

    {{x0, _, _}, {x1, _, _}} =
      create_one_circuit(coords_and_dists, length(coords), MapSet.new())

    x0 * x1
  end

  # ================ helpers ================

  # Recursively connects coords and add to the proper circuits, creating and
  # merging circuits when necessary. Returns the resulting circuits after
  # `num_iters`.
  defp create_circuits(_, 0, circuits), do: MapSet.to_list(circuits)

  defp create_circuits([{c0, c1, _} | coords_and_dists], num_iters, circuits) do
    create_circuits(coords_and_dists, num_iters - 1, add_coords_to_circuits(c0, c1, circuits))
  end

  # Recursively connects coords and add to the proper circuits, creating and
  # merging circuits when necessary. When we finally create one circuit
  # containing all the coordinates, return the last two coords that made
  # that happen.
  defp create_one_circuit([{c0, c1, _} | coords_and_dists], num_coords, circuits) do
    circuits = add_coords_to_circuits(c0, c1, circuits)

    if MapSet.size(circuits) == 1 && MapSet.size(hd(MapSet.to_list(circuits))) == num_coords do
      {c0, c1}
    else
      create_one_circuit(coords_and_dists, num_coords, circuits)
    end
  end

  # Given two coordinates and a list of all existing circuits, "connects"
  # the two coordinates by placing them in the proper circuits, possibly
  # creating a new circuit. Returns the new list of circuits.
  defp add_coords_to_circuits(c0, c1, circuits) do
    c0_circuit = Enum.find(circuits, fn circuit -> MapSet.member?(circuit, c0) end)
    c1_circuit = Enum.find(circuits, fn circuit -> MapSet.member?(circuit, c1) end)

    case {c0_circuit, c1_circuit} do
      # create a new circuit
      {nil, nil} ->
        circuits |> MapSet.put(MapSet.new([c0, c1]))

      # add c0 to c1's circuit
      {nil, c1c} ->
        c1c = MapSet.put(c1c, c0)
        circuits |> MapSet.put(c1c) |> MapSet.delete(c1_circuit)

      # add c1 to c0's circuit
      {c0c, nil} ->
        c0c = MapSet.put(c0c, c1)
        circuits |> MapSet.put(c0c) |> MapSet.delete(c0_circuit)

      # both already in a circuit together
      {cc, cc} ->
        circuits

      # merge the two circuits into one
      {c0c, c1c} ->
        circuits
        |> MapSet.put(MapSet.union(c0c, c1c))
        |> MapSet.delete(c0_circuit)
        |> MapSet.delete(c1_circuit)
    end
  end

  # Returns a list of all combinations of coords as {coord, coord,
  # squared_distance} tuples.
  defp pairs_and_distances_between(coords) do
    coords
    |> CE.combinations(2)
    |> Enum.map(fn [c0, c1] -> {c0, c1, squared_dist(c0, c1)} end)
  end

  # Returns the squared distance between two coords.
  defp squared_dist({x0, y0, z0}, {x1, y1, z1}) do
    dx = abs(x1 - x0)
    dy = abs(y1 - y0)
    dz = abs(z1 - z0)
    dx * dx + dy * dy + dz * dz
  end

  # Returns a list of {x, y, z} tuples.
  defp parse_3d(lines) do
    lines |> Enum.map(&line_to_coords_tuple/1)
  end

  # Converts a single line into an {x, y, z} tuple.
  defp line_to_coords_tuple(line) do
    line
    |> String.split(",")
    |> Enum.map(&String.to_integer/1)
    |> List.to_tuple()
  end
end
